// langbly/dist/cjs/client.js
"use strict";
/**
 * Langbly — Official JavaScript/TypeScript SDK for the Langbly translation API.
 *
 * A drop-in replacement for Google Translate v2 — powered by LLMs.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Langbly = exports.AuthenticationError = exports.RateLimitError = exports.LangblyError = void 0;
class LangblyError extends Error {
    constructor(message, status = 0, code = "") {
        super(message);
        this.name = "LangblyError";
        this.status = status;
        this.code = code;
    }
}
exports.LangblyError = LangblyError;
class RateLimitError extends LangblyError {
    constructor(message, retryAfter = null) {
        super(message, 429, "RATE_LIMITED");
        this.name = "RateLimitError";
        this.retryAfter = retryAfter;
    }
}
exports.RateLimitError = RateLimitError;
class AuthenticationError extends LangblyError {
    constructor(message) {
        super(message, 401, "UNAUTHENTICATED");
        this.name = "AuthenticationError";
    }
}
exports.AuthenticationError = AuthenticationError;
const RETRIABLE_STATUS_CODES = new Set([429, 500, 502, 503, 504]);
class Langbly {
    constructor(options) {
        if (!options.apiKey) {
            throw new Error("apiKey is required");
        }
        this.apiKey = options.apiKey;
        this.baseUrl = (options.baseUrl ?? "https://api.langbly.com").replace(/\/$/, "");
        this.timeout = options.timeout ?? 30000;
        this.maxRetries = options.maxRetries ?? 2;
    }
    async translate(text, options) {
        const q = Array.isArray(text) ? text : [text];
        const body = { q, target: options.target };
        if (options.source)
            body.source = options.source;
        if (options.format)
            body.format = options.format;
        if (options.context)
            body.context = options.context;
        if (options.instructions)
            body.instructions = options.instructions;
        if (options.glossary)
            body.glossary = options.glossary;
        const data = await this.postWithRetry("/language/translate/v2", body);
        const translations = data.data.translations.map((item) => ({
            text: item.translatedText,
            source: item.detectedSourceLanguage ?? options.source ?? "",
            model: item.model,
        }));
        return Array.isArray(text) ? translations : translations[0];
    }
    /**
     * Detect the language of text.
     */
    async detect(text) {
        const data = await this.postWithRetry("/language/translate/v2/detect", {
            q: text,
        });
        const det = data.data.detections[0][0];
        return {
            language: det.language,
            confidence: det.confidence ?? 0,
        };
    }
    /**
     * List supported languages.
     */
    async languages(options) {
        const params = new URLSearchParams();
        if (options?.target)
            params.set("target", options.target);
        const qs = params.toString();
        const url = `/language/translate/v2/languages${qs ? `?${qs}` : ""}`;
        const resp = await this.fetchWithRetry(`${this.baseUrl}${url}`, {
            method: "GET",
        });
        const data = await resp.json();
        return data.data.languages.map((lang) => ({
            code: lang.language,
            name: lang.name,
        }));
    }
    async postWithRetry(path, body) {
        const resp = await this.fetchWithRetry(`${this.baseUrl}${path}`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(body),
        });
        return resp.json();
    }
    async fetchWithRetry(url, init) {
        const headers = {
            Authorization: `Bearer ${this.apiKey}`,
            "User-Agent": "langbly-js/0.1.0",
            ...init.headers,
        };
        let lastError;
        for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
            let resp;
            try {
                resp = await fetch(url, {
                    ...init,
                    headers,
                    signal: AbortSignal.timeout(this.timeout),
                });
            }
            catch (err) {
                // Network error or timeout
                lastError = err instanceof Error ? err : new Error(String(err));
                if (attempt < this.maxRetries) {
                    await this.sleep(this.backoffDelay(attempt));
                    continue;
                }
                throw new LangblyError(`Request failed after ${this.maxRetries + 1} attempts: ${lastError.message}`, 0, "CONNECTION_ERROR");
            }
            if (resp.ok) {
                return resp;
            }
            // Non-retriable error — throw immediately
            if (!RETRIABLE_STATUS_CODES.has(resp.status)) {
                await this.throwForStatus(resp);
            }
            // Retriable error — retry if we have attempts left
            if (attempt < this.maxRetries) {
                const delay = this.getRetryDelay(resp, attempt);
                await this.sleep(delay);
                continue;
            }
            // Final attempt failed
            await this.throwForStatus(resp);
        }
        // Should not reach here
        throw lastError ?? new LangblyError("Request failed");
    }
    async throwForStatus(resp) {
        let message = resp.statusText;
        let code = "";
        try {
            const err = await resp.json();
            message = err?.error?.message ?? message;
            code = err?.error?.status ?? "";
        }
        catch {
            // ignore parse errors
        }
        if (resp.status === 401) {
            throw new AuthenticationError(message);
        }
        if (resp.status === 429) {
            const retryAfter = this.parseRetryAfter(resp);
            throw new RateLimitError(message, retryAfter);
        }
        throw new LangblyError(message, resp.status, code);
    }
    parseRetryAfter(resp) {
        const header = resp.headers.get("retry-after");
        if (!header)
            return null;
        const value = Number(header);
        return Number.isFinite(value) ? value : null;
    }
    getRetryDelay(resp, attempt) {
        const retryAfter = resp.headers.get("retry-after");
        if (retryAfter) {
            const value = Number(retryAfter);
            if (Number.isFinite(value)) {
                return Math.min(value * 1000, 30000);
            }
        }
        return Math.min(500 * 2 ** attempt, 10000);
    }
    backoffDelay(attempt) {
        return Math.min(500 * 2 ** attempt, 10000);
    }
    sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
}
exports.Langbly = Langbly;
